mcdc/loop.py

```diff
@@ -149,6 +178,155 @@ def loop_eigenvalue(mcdc):
+def prep_particle(P, prog):
+    mcdc = adapt.device(prog)
+
+    # Apply weight window
+    if mcdc["technique"]["weight_window"]:
+        kernel.weight_window(P, prog)
+
+    # Particle tracker
+    if mcdc["setting"]["track_particle"]:
+        kernel.allocate_pid(P, mcdc)
+
+
+@njit(cache=caching)
+def exhaust_active_bank(prog):
+    mcdc = adapt.device(prog)
+    P = adapt.local_particle()
+    # Loop until active bank is exhausted
+    while kernel.get_bank_size(mcdc["bank_active"]) > 0:
+        # Get particle from active bank
+        kernel.get_particle(P, mcdc["bank_active"], mcdc)
+
+        prep_particle(P, prog)
+
+        # Particle loop
+        loop_particle(P, mcdc)
+
+
+@njit(cache=caching)
+def source_closeout(prog, idx_work, N_prog):
+    mcdc = adapt.device(prog)
+
+    # Tally history closeout for one-batch fixed-source simulation
+    if not mcdc["setting"]["mode_eigenvalue"] and mcdc["setting"]["N_batch"] == 1:
+        kernel.tally_closeout_history(mcdc)
+
+    # Tally history closeout for multi-batch uq simulation
+    if mcdc["technique"]["uq"]:
+        kernel.uq_tally_closeout_history(mcdc)
+
+    # Progress printout
+    percent = (idx_work + 1.0) / mcdc["mpi_work_size"]
+    if mcdc["setting"]["progress_bar"] and int(percent * 100.0) > N_prog:
+        N_prog += 1
+        with objmode():
+            print_progress(percent, mcdc)
+
+
+@njit(cache=caching)
+def source_dd_resolution(prog):
+    mcdc = adapt.device(prog)
+
+    kernel.dd_particle_send(mcdc)
+    terminated = False
+    max_work = 1
+    kernel.dd_recv(mcdc)
+    if mcdc["domain_decomp"]["work_done"]:
+        terminated = True
+
+    while not terminated:
+        if kernel.get_bank_size(mcdc["bank_active"]) > 0:
+            P = adapt.local_particle()
+            # Loop until active bank is exhausted
+            while kernel.get_bank_size(mcdc["bank_active"]) > 0:
+
+                kernel.get_particle(P, mcdc["bank_active"], mcdc)
+                if not kernel.particle_in_domain(P, mcdc) and P["alive"] == True:
+                    print(f"recieved particle not in domain")
+
+                # Apply weight window
+                if mcdc["technique"]["weight_window"]:
+                    kernel.weight_window(P, mcdc)
+
+                # Particle tracker
+                if mcdc["setting"]["track_particle"]:
+                    mcdc["particle_track_particle_ID"] += 1
+
+                # Particle loop
+                loop_particle(P, mcdc)
+
+                # Tally history closeout for one-batch fixed-source simulation
+                if (
+                    not mcdc["setting"]["mode_eigenvalue"]
+                    and mcdc["setting"]["N_batch"] == 1
+                ):
+                    kernel.tally_closeout_history(mcdc)
+
+        # Send all domain particle banks
+        kernel.dd_particle_send(mcdc)
+
+        kernel.dd_recv(mcdc)
+
+        # Progress printout
+        """
+        percent = 1 - work_remaining / max_work
```

Comment:
Should this be under an input toggle? Maybe we can declare verbosity

Reply:
It can be, if we wish.

This snippet is a holdover from pre-revamp domain decomposition, so there's no overarching intent to its inclusion.

Reply:
I've just looked into this a bit more. It looks like the refactor makes finding an estimation of progress more difficult. Since each rank is only notified of work that it has been assigned  or whether a halting condition has been reached, it is not currently possible to reliably estimate the amount of work that has been completed. This functionality could be added in later, but it is probably beyond the scope of this PR.